### PR TITLE
Remove pkg/client, use controller-runtime Client in clusterctl

### PR DESCRIPTION
**Pull request by vincepri**
**What this PR does / why we need it**:
This PR is a first step to use controller-runtime `Client` and remove the clientset from Cluster API.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #1240

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go [hunk 1]
```diff
@@ -486,180 +493,81 @@ func (c *client) CreateMachines(machines []*clusterv1.Machine, namespace string)
...
-		clusters, err := c.clientSet.ClusterV1alpha2().Clusters("").List(metav1.ListOptions{})
-		if err != nil {
-			return errors.Wrap(err, "error listing Clusters in all namespaces")
-		}
-		for _, cluster := range clusters.Items {
-			if _, ok := seen[cluster.Namespace]; !ok {
-				seen[cluster.Namespace] = true
-			}
-		}
+	if err := c.clientSet.DeleteAllOf(ctx, &clusterv1.Cluster{}, ctrlclient.InNamespace(namespace)); err != nil {
```
[the remark this answers is not in the document]
Comment (vincepri) [reply cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go]: The comment reads `// DeleteAllOf deletes all objects of the given type matching the given options.`, which should be the case with an empty namespace. That said, we'll definitely need to test this later on, maybe rewriting this whole section

File: cmd/clusterctl/clientcmd/configutil.go [hunk 2]
```diff
@@ -45,12 +45,13 @@ func NewCoreClientSetForDefaultSearchPath(kubeconfigPath string, overrides clien
 // NewClusterAPIClientForDefaultSearchPath creates a Cluster API clientset. If the kubeconfigPath is specified then the configuration is loaded from that path.
```
[the remark this answers is not in the document]
Comment (vincepri) [reply cmd/clusterctl/clientcmd/configutil.go]: Updated! Thanks for catching that :)

**Comment by vincepri** [comment]
@detiber This should be ready for review now 🙂

File: Makefile [hunk 3]
```diff
@@ -98,36 +95,14 @@ lint-full: ## Run slower linters to detect possible issues
...
-		--go-header-file=./hack/boilerplate/boilerplate.generatego.txt
-	bin/informer-gen \
-		--input-dirs sigs.k8s.io/cluster-api/pkg/apis/deprecated/v1alpha1,sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha2 \
-		--versioned-clientset-package sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset \
-		--listers-package sigs.k8s.io/cluster-api/pkg/client/listers_generated \
-		--output-package sigs.k8s.io/cluster-api/pkg/client/informers_generated \
-		--go-header-file=./hack/boilerplate/boilerplate.generatego.txt
+.PHONY: generate-deepcopy
+generate-deepcopy: ## Runs controller-gen to generate deepcopy files
+	go run vendor/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go \
```
Comment (detiber) [inline Makefile]: Should we handle this similar to how we've been handling controller-gen? https://github.com/kubernetes-sigs/cluster-api-provider-docker/blob/master/Makefile#L78-L80

File: cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go [hunk 4]
```diff
@@ -727,75 +639,75 @@ func (c *client) WaitForResourceStatuses() error {
...
-		if err != nil {
-			return false, nil
-		}
-		if response != nil {
-			return false, nil
-		}
-		return true, nil
-	})
-}
+// func (c *client) waitForClusterDelete(namespace string) error {
```
Comment (detiber) [inline cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go]: Might as well delete these, can always refer back to the `release-0.1` branch if we need to bring them back.

File: cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go [hunk 5]
```diff
@@ -38,25 +39,28 @@ import (
...
 )
 
 const (
-	retryIntervalKubectlApply   = 10 * time.Second
-	retryIntervalResourceReady  = 10 * time.Second
-	retryIntervalResourceDelete = 10 * time.Second
-	timeoutKubectlApply         = 15 * time.Minute
-	timeoutResourceReady        = 15 * time.Minute
-	timeoutMachineReady         = 30 * time.Minute
-	timeoutResourceDelete       = 15 * time.Minute
```
[the remark this answers is not in the document]
Comment (vincepri) [reply cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go]: We no longer wait for deletion requests to complete as we used to before

File: Makefile (hunk 3 above)
Comment (vincepri) [reply Makefile]: Definitely! I'll follow-up once we move CAPA to having a hack submodule as well

File: cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go [hunk 6]
```diff
@@ -860,32 +772,33 @@ func (c *client) waitForKubectlApply(manifest string) error {
...
+	// 	_, err := cs.Discovery().ServerResourcesForGroupVersion("cluster.k8s.io/v1alpha2")
+	// 	if err == nil {
+	// 		return true, nil
+	// 	}
+	// 	return false, nil
+	// })
+
+	// if err != nil {
+	// 	return err
+	// }
```
[the remark this answers is not in the document]
Comment (vincepri) [reply cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go]: Done :)

File: cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go [hunk 7]
```diff
@@ -246,52 +250,55 @@ func (c *client) GetCluster(name, ns string) (*clusterv1.Cluster, error) {
...
 	}
 
 	return nil
 }
 
-func (c *client) GetClusters(namespace string) ([]*clusterv1.Cluster, error) {
-	clusters := []*clusterv1.Cluster{}
-	clusterlist, err := c.clientSet.ClusterV1alpha2().Clusters(namespace).List(metav1.ListOptions{})
-	if err != nil {
+func (c *client) GetClusters(namespace string) (items []*clusterv1.Cluster, _ error) {
```
[the remark this answers is not in the document]
Comment (vincepri) [reply cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go]: Fixed

File: cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (detiber) [reply cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go]: I'm wondering if that will cause us issues, but we can cross that bridge when we get there.

**Comment by detiber** [comment]
/lgtm
let's go ahead and get this merged and iterate on it later as needed.

**Pull request merged**